File: src/main/java/org/sonar/plugins/xquery/parser/reporter/ProblemReporter.java
/*
 * © 2014 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.parser.reporter;

import org.antlr.runtime.Token;

import java.util.ArrayList;
import java.util.List;

public class ProblemReporter {
    
    private boolean failOnError;
    private boolean outputError;

    private List<Problem> problems;

    public ProblemReporter() {
        this(false);
    }

    public ProblemReporter(boolean failOnError) {
        this.failOnError = failOnError;
        this.outputError = true;
        problems = new ArrayList<Problem>();
    }

    public List<Problem> getProblems() {
        return problems;
    }

    public boolean isFailOnError() {
        return failOnError;
    }
    
    public boolean isOutputError() {
        return outputError;
    }
    
    public void reportError(String id, String message, Token token) {
        Problem problem = new Problem(id, message, token);
        problems.add(problem);
        if (failOnError) {
            throw new RuntimeException(problem.getMessageString());
        } else if (outputError) {
            System.err.println(problem.getId() + problem.getMessageString());
        }
    }
        
    public void setFailOnError(boolean failOnError) {
        this.failOnError = failOnError;
    }
    
    public void setOutputError(boolean outputError) {
        this.outputError = outputError;
    }
}

File: src/main/java/org/sonar/plugins/xquery/parser/reporter/ProblemReporter.kt
/*
 * © 2014 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.parser.reporter

import org.antlr.v4.runtime.BaseErrorListener
import org.antlr.v4.runtime.RecognitionException
import org.antlr.v4.runtime.Recognizer
import org.antlr.v4.runtime.misc.ParseCancellationException
import java.util.*

class ProblemReporter(var isFailOnError: Boolean = false) : BaseErrorListener() {
    var isOutputError: Boolean = false

    val problems: MutableList<Problem>

    init {
        this.isOutputError = true
        problems = ArrayList()
    }

    override fun syntaxError(recognizer: Recognizer<*, *>?, offendingSymbol: Any, line: Int, charPositionInLine: Int, msg: String, e: RecognitionException?) {
        // Bug with EOF?  Even when adding it to the grammar implicitly it still doesn't match right
        if (!msg.startsWith("mismatched input '<EOF>'")) {
            val problem = Problem(line, charPositionInLine, msg)
            problems.add(problem)
            if (isFailOnError) {
                throw ParseCancellationException(problem.getMessageString())
            } else if (isOutputError) {
                System.err.println(problem.getMessageString())
            }
        }
    }
}
